# 지역 변수 vs 전역 변수
## Developer / Analyst / Project Manager 관점

> **핵심 비유:** 변수가 보는 범위(**scope**)가 곧 각 역할이 책임지고 이해하는 **프로젝트 범위**다.
- **지역 변수(Local):** 함수 안에서 정의 → 함수 안에서만 존재 (밖에서 접근 불가)
- **전역 변수(Global):** 함수 밖에서 선언 → 프로그램 어디서나 접근 가능


---
## 1. 아주 간단한 예제 (기본 개념)

**① 지역 변수 — 함수 안에서만 존재**

In [1]:
def greet():
    msg = "함수 안에서 안녕!"   # 지역 변수
    print(msg)

greet()
# print(msg)  # NameError: msg는 함수 밖에서 안 보임

함수 안에서 안녕!


**② 전역 변수 — 어디서나 접근 가능**

In [2]:
msg = "Python is awesome!"   # 전역 변수

def display():
    print("함수 안:", msg)       # 전역 변수 읽기 OK

display()
print("함수 밖:", msg)            # 함수 밖에서도 OK

함수 안: Python is awesome!
함수 밖: Python is awesome!


**③ Shadowing — 같은 이름의 지역 변수가 전역을 가림**

In [ ]:
s = "I love Yeardream"   # 전역

def fun():
    s = "Me too."              # 지역(전역을 가림)
    print(s)

fun()        # → Me too.
print(s)     # → I love GeeksforGeeks (전역은 그대로)

Me too.
I love GeeksforGeeks


**④ 전역 변수를 함수 안에서 *수정* 하려면 `global` 필요**

In [4]:
count = 0

def increase():
    global count   # 이 선언이 있어야 전역을 바꿀 수 있음
    count += 1

increase()
increase()
print(count)   # → 2

2


---
## 2. 프로젝트 시나리오로 변형

같은 개념을 "프로젝트 예산/진행 상황"에 빗대어 봅니다.
변수의 scope = 각 역할이 다루는 범위.

In [5]:
# 전역 변수 (Global) — 프로젝트 전체가 공유하는 정보
project_budget = 100_000_000   # 전체 예산
project_status = "진행중"       # 전체 상태

# ── Developer 영역 ──
def develop_feature():
    local_task = "로그인 기능 구현"   # 지역: 이 함수(개발자) 안에서만 존재
    hours_spent = 8                  # 지역
    print(f"[개발자] 작업: {local_task}, 투입시간: {hours_spent}h")
    print(f"[개발자] (참고) 전체 상태: {project_status}")  # 전역은 읽기 가능

develop_feature()
# print(local_task)  # NameError! 개발자의 local_task는 밖에서 안 보임

# ── Analyst 영역 ──
def analyze():
    module_a_cost = 30_000_000       # 지역: 분석 범위 안에서만 의미 있음
    module_b_cost = 25_000_000
    used_ratio = (module_a_cost + module_b_cost) / project_budget  # 전역 참조해 분석
    print(f"[분석가] 예산 소진율: {used_ratio:.0%}")

analyze()

# ── Project Manager 영역 ──
def manage():
    global project_status            # 전역 변수를 '수정'할 권한
    project_status = "위험"           # 변경이 프로젝트 전체에 반영됨
    print(f"[PM] 상태 변경 -> {project_status}")

manage()
print(f"최종 전체 상태: {project_status}")   # PM의 변경이 전역에 남음

[개발자] 작업: 로그인 기능 구현, 투입시간: 8h
[개발자] (참고) 전체 상태: 진행중
[분석가] 예산 소진율: 55%
[PM] 상태 변경 -> 위험
최종 전체 상태: 위험


---
## 3. 역할별 관점 정리 (프로젝트 이해도 = scope)

### Developer — 지역 변수 중심
- 자기 **함수(모듈/기능)** 안에서 작동
- `local_task`, `hours_spent`처럼 **내 작업 단위**에서만 의미 있는 값을 다룸
- 내 함수 내부 변수는 밖으로 안 나감 → 다른 사람이 직접 못 건드림
- 전역 상태는 **읽어서 참고**만 함
- 이해의 **넓이**는 부분적(local)이지만, 자기 영역에선 **가장 깊음**

### Analyst — 지역 ↔ 전역을 연결
- 여러 모듈 값을 모아(`module_a_cost`, `module_b_cost`) **전역 맥락**(`project_budget`)과 연결
- 지역 디테일 + 전역 맥락을 **잇는** 역할
- 데이터를 바꾸지는 않음(전역 **수정 X**), "예산 55% 소진" 같은 **해석**을 생성
- 개발자보다 넓게 보되, PM 같은 변경 권한은 없는 **중간 scope**

### Project Manager — 전역 변수 중심
- `global` 키워드를 쓰는 **유일한 역할** = 전체에 영향 주는 값 **변경 권한**
- `project_status`를 바꾸면 함수가 끝나도 전역에 남아 **모두에게 전파**
- 개별 `local_task` 같은 세부는 직접 안 봄 (전체 namespace를 봄)
- 이해의 **넓이가 가장 큼**, 단 특정 기능의 **깊이는 개발자에게 위임**


---
## 4. 한눈에 비교

| 관점 | 변수 비유 | scope(이해 범위) | 권한 |
|------|-----------|------------------|------|
| Developer | Local | 자기 함수(모듈) 안에서 깊게 | 지역만 자유, 전역은 읽기 |
| Analyst | Local → Global 연결 | 여러 모듈 + 전체 맥락 | 읽고 해석, 수정 X |
| PM | Global | 프로젝트 전체를 넓게 | `global`로 전체 상태 변경 |

### Shadowing의 현실 비유 (용어 혼선)
- 개발자의 "완료"(내 기능 기준)과 PM의 "완료"(릴리스 기준)는 **같은 단어, 다른 scope**
- 같은 이름의 지역 변수가 전역을 가리듯, 같은 용어가 역할마다 다른 값을 가리킴
- → 소통 시 **어느 scope의 값을 말하는지** 맞추는 것이 중요
